x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/update/update_gaps.ts

```diff
@@ -0,0 +1,219 @@
+
+const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+const updateSingleGap = async (
+  gap: Gap,
+  {
+    backfillSchedule,
+    savedObjectsRepository,
+    shouldRefetchAllBackfills,
+    backfillClient,
+    actionsClient,
+    alertingEventLogger,
+    logger,
+    ruleId,
+    eventLogClient,
+  }: {
+    backfillSchedule?: BackfillSchedule[];
+    savedObjectsRepository: ISavedObjectsRepository;
+    shouldRefetchAllBackfills?: boolean;
+    backfillClient: BackfillClient;
+    actionsClient: ActionsClient;
+    alertingEventLogger: AlertingEventLogger;
+    logger: Logger;
+    ruleId: string;
+    eventLogClient: IEventLogClient;
+  }
+) => {
+  let currentGap = gap;
+  let retryCount = 0;
+
+  while (retryCount <= MAX_RETRIES) {
+    try {
+      // If we're retrying, refetch the latest gap state
+      if (retryCount > 0 && currentGap.internalFields) {
+        const updatedGap = await findGapById({
+          eventLogClient,
+          logger,
+          params: {
+            gapId: currentGap.internalFields._id,
+            ruleId,
+          },
+        });
+
+        if (!updatedGap) {
+          logger.warn(
+            `Gap ${currentGap.internalFields._id} not found during retry, it was removed by another process`
+          );
+          return;
+        }
+        currentGap = updatedGap;
+      }
+
+      const hasFailedBackfillTask = backfillSchedule?.some(
+        (scheduleItem) =>
+          scheduleItem.status === adHocRunStatus.ERROR ||
+          scheduleItem.status === adHocRunStatus.TIMEOUT
+      );
+
+      if (backfillSchedule && !hasFailedBackfillTask) {
+        updateGapFromSchedule({
+          gap: currentGap,
+          backfillSchedule,
+        });
+      }
+
+      if (hasFailedBackfillTask || !backfillSchedule || shouldRefetchAllBackfills) {
+        await calculateGapStateFromAllBackfills({
+          gap: currentGap,
+          savedObjectsRepository,
+          ruleId,
+          backfillClient,
+          actionsClient,
+        });
+      }
+
+      const esGap = currentGap.getEsObject();
+      const internalFields = currentGap.internalFields;
+
+      if (internalFields) {
+        await alertingEventLogger.updateGap({
+          internalFields,
+          gap: esGap,
+        });
+      }
+      return;
+    } catch (e) {
+      if (e.statusCode === CONFLICT_STATUS_CODE) {
+        retryCount++;
+        if (retryCount > MAX_RETRIES) {
+          logger.error(
+            `Failed to update gap: ${currentGap?.internalFields?._id} after ${MAX_RETRIES} retries due to conflicts`
+          );
+          throw e;
+        }
+        // Use exponential backoff with some randomness: 2s, 4s, 8s
+        const retryDelaySec = Math.min(Math.pow(2, retryCount), 30);
+        await delay(retryDelaySec * 1000 * Math.random());
+        continue;
+      }
+      throw e;
+    }
+  }
+};
+
+export const updateGaps = async (params: UpdateGapsParams) => {
```

Comment:
are we able to run a test where the number of gaps we have to update is very large with a backfill with a large number of schedule entries (like a 24 hour ourage with 5m rule increments) and see how long it takes to perform these calculations (not including the cost of performing the ES updates)?

Reply:
This is results I got for 500/1000 gaps. As we process it now in batch of 500 gaps,

5 minute rule interval for 1 day of gaps it's - 5*12*24 = 1440

Test results for only processing, no es requests
500 gaps, 1000 schedules  = 300ms
500 gaps, 10.000 schedules = 2846ms
1000 gaps, 10.000 schedules = 5827ms 